server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -1010,6 +1017,89 @@ public boolean isForceExecution() {
         );
     }
 
+    /**
+     * Determines if an index name is associated with either an existing data stream or a template
+     * for one that has the failure store enabled.
+     * @param indexName The index name to check.
+     * @param metadata Cluster state metadata.
+     * @param epochMillis A timestamp to use when resolving date math in the index name.
+     * @return true if the given index name corresponds to a data stream with a failure store,
+     * or if it matches a template that has a data stream failure store enabled.
+     */
+    static boolean shouldStoreFailure(String indexName, Metadata metadata, long epochMillis) {
+        return DataStream.isFailureStoreEnabled()
+            && resolveFailureStoreFromMetadata(indexName, metadata, epochMillis).or(
+                () -> resolveFailureStoreFromTemplate(indexName, metadata)
+            ).orElse(false);
+    }
+
+    /**
+     * Determines if an index name is associated with an existing data stream that has a failure store enabled.
+     * @param indexName The index name to check.
+     * @param metadata Cluster state metadata.
+     * @param epochMillis A timestamp to use when resolving date math in the index name.
+     * @return true if the given index name corresponds to an existing data stream with a failure store enabled.
+     */
+    private static Optional<Boolean> resolveFailureStoreFromMetadata(String indexName, Metadata metadata, long epochMillis) {
```

Comment:
I think we could add a unit test for this method, which could be useful. What do you think?